docker-app/qfieldcloud/core/utils2/storage.py

```diff
@@ -96,3 +100,54 @@ def remove_project_thumbail(project: "Project") -> None:  # noqa: F821
     bucket = qfieldcloud.core.utils.get_s3_bucket()
     key = project.thumbnail_uri
     bucket.object_versions.filter(Prefix=key).delete()
+
+
+def purge_old_file_versions(project: "Project") -> None:  # noqa: F821
+    """
+    Deletes old versions of all files in the given project. Will keep __3__
+    versions for COMMUNITY user accounts, and __10__ versions for PRO user
+    accounts
+    """
+
+    logger.info(f"Cleaning up old files for {project}")
+
+    # Determine account type
+    account_type = project.owner.useraccount.account_type
+    if account_type == UserAccount.TYPE_COMMUNITY:
+        keep_count = 3
+    elif account_type == UserAccount.TYPE_PRO:
+        keep_count = 10
+    else:
+        raise NotImplementedError(f"Unknown account type {account_type}")
+
+    logger.debug(f"Keeping {keep_count} versions")
+
+    # Process file by file
+    for file in qfieldcloud.core.utils.get_project_files_with_versions(project.pk):
+
+        filename = file.latest.name
+        old_versions = file.versions
+
+        # Sort by date (newest first)
+        old_versions.sort(key=lambda i: i.last_modified, reverse=True)
+
+        # Skip the newest N
+        old_versions_to_purge = old_versions[keep_count:]
```

Comment:
```suggestion
        old_versions_to_purge = sorted(file.versions, key=lambda i: i.last_modified)[-keep_count:]
```

Why I would prefer it:
- we make a copy of `file.versions`, instead of sorting in place,  which might affect the code in the future if the versions are cached.
- no need for the extra var `old_versions`

I would even move the sorting functionality directly to `S3ObjectWithVersions`. TBH the current QFC code already assumes that the versions are sorted chronologically by `boto3`, but I do agree and extra sorting would not hurt.

Reply:
Ok changed so that it doesn't sort in place (didn't exactly take your rewrite as it changed the logic)

Yes indeed I saw that get_project_files_with_versions expects already sorted versions from boto3 (not sure boto3 guarantees that though ? may depend on the provider ?)

Anyway, since it's about deleting stuff, better be extra careful here.

Reply:
yes, extra care makes sense.